python/lib/core/dmod/core/context.py

```diff
@@ -0,0 +1,607 @@
+
+    return found_item
+
+
+def member_should_be_exposed_to_proxy(member: typing.Any) -> bool:
+    """
+    Determine whether the member of a class should be exposed through a proxy
+
+    Args:
+        member: The member of a class that might be exposed
+
+    Returns:
+        True if the member should be accessible via the proxy
+    """
+    if inspect.isclass(member) or inspect.ismodule(member):
+        return False
+
+    if isinstance(member, property):
+        return True
+
+    member_is_callable = inspect.isfunction(member) or inspect.ismethod(member) or inspect.iscoroutinefunction(member)
+    if not member_is_callable:
+        return False
+
+    member_name = getattr(member, "__name__", None)
+
+    # Not having a name is not a disqualifier.
+    # We want to include properties in this context and they won't have names here
+    if member_name is None:
+        return False
+
+    # Double underscore functions/attributes (dunders in pythonic terms) are denoted by '__xxx__'
+    # and are special functions that define things like behavior of `instance[key]`, `instance > other`,
+    # etc. Only SOME of these are valid, so we need to ensure that these fall into the correct subset
+    member_is_dunder = member_name.startswith("__") and member_name.endswith("__")
+
+    if member_is_dunder:
+        return member_name in __ACCEPTABLE_DUNDERS
+
+    # A member is considered private if the name is preceded by '_'. Since these are private,
+    # they shouldn't be used by outside entities, so we'll leave these out
+    if member_name.startswith("_"):
+        return False
+
+    return True
+
+
+def make_proxy_type(
+    cls: typing.Type,
+    exposure_criteria: typing.Callable[[typing.Any], bool] = None
+) -> TypeOfRemoteObject:
+    """
+    Create a remote interface class with the given name and with the list of names of functions that may be
+    called which will call the named functions on the remote object
+
+    Args:
+        cls: The class to create a proxy for
+        exposure_criteria: A function that will decide if a bound object should be exposed through the proxy
+
+    Returns:
+        A proxy type that can be used to interact with the object instantiated in the manager process
+    """
+    if exposure_criteria is None:
+        exposure_criteria = member_should_be_exposed_to_proxy
+
+    logging.debug(f"Creating a proxy class for {cls.__name__} in process {os.getpid()}")
+
+    # This dictionary will contain references to functions that will be placed in a dynamically generated proxy class
+    new_class_members: typing.Dict[str, typing.Union[typing.Dict, typing.Callable, typing.Tuple]] = {}
+
+    # Determine what members and their names to expose based on the passed in criteria for what is valid to expose
+    members_to_expose = dict(inspect.getmembers(object=cls, predicate=exposure_criteria))
+    lines_of_code: typing.List[str] = []
+    for member_name, member in members_to_expose.items():
```

Comment:
yeah, it's due to some of the weird behavior of the `type` constructor. You end up with scoping issues if the functions aren't defined via `exec`. Calling `exec` makes me want to hurl, but python does that because it allows the separation of scope which is defined/encapsulated by `new_class_members`.